components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3188,39 +3215,68 @@ subroutine evaporate_sublimate_precip(qr_incld,qc_incld,nr_incld,qitot_incld,
-   ! been removed from the original P3-WRF.
-
-   ! Determine temporary cloud fraction, set to zero if cloud water + ice is
-   ! very small.  This will ensure that evap/subl of precip occurs over entire
-   ! grid cell, since min cloud fraction is specified otherwise.
-   if (qc_incld + qitot_incld < 1.e-6_rtype) then
-      cld = 0._rtype
+   ! First, check that qv<=qvs. If not, unexpected supersaturation has occurred
+   ! and the assumptions of this parameterization will have been violated.
+   ! This check should eventually be commented out and/or moved to a debug-mode
+   ! feature.
+   !if ( qv-qvs .gt. 1.e-12_rtype ) then
+   !    print*
+   !    print*,'In evap_precip, qv supersaturated. qv=',qv,', qvs=',qvs,', diff=',qv-qvs
+   !    print*
+   !    call endscreamrun("qv supersaturated in evap_precip")
+   ! endif
+   
+   ! Determine temporary cloud fraction which ramps to 0 as cloud water
+   ! approaches zero. This avoids problems due to lcldm having a positive
+   ! minimum value which would result in evaporation never occurring in a
+   ! small portion of the grid. Ramping rather than thresholding is used
+   ! to preserve convergence.
+   if (qc_incld < 1.e-6_rtype) then
+      cld = lcldm*qc_incld/1.e-6_rtype
    else
       cld = lcldm
    end if
 
    ! Only calculate if there is some rain fraction > cloud fraction
+   ! TODO: discontinuous behavior if qr_incld.ge.qsmall may cause
+   !       convergence issues. Treatment should be improved.
    qrevp = 0.0_rtype
-   if (rcldm > cld) then
-      ! calculate q for out-of-cloud region
+   nrevp = 0.0_rtype
+   if (rcldm > cld .and. qr_incld.ge.qsmall) then
+      
+      ! calculate q for clear-sky region
+      ! TODO: as cld -> 1, qclr could blow up. Fixing this requires knowing
+      !       what qclr should do in this limit, which requires L'Hospital's
+      !       rule and an understanding of how qv approaches qvs as a function
+      !       of cld. This relationship will vary depending on your cloud macrophysics
+      !       scheme and any microphysical processes applied before now. Practically,
+      !       we haven't seen the following line to crash the code, so leaving as-is
+      !       with limiters to prevent evap from getting too out of control. A real fix
+      !       may require a completely different approach to rain evaporation.
       qclr = (qv-cld*qvs)/(1._rtype-cld)
+      qclr = max(0._rtype, qclr) !as cld approaches 1, qclr goes negative.
+      qclr = min(qclr,qv) !qv<qvs check above should prevent this, but just in case...
 
-      ! rain evaporation
-      if (qr_incld.ge.qsmall) then
-         qrevp = epsr * (qclr-qvs)/ab
-      end if
+      ! compute in-evaporating-region evap rate
+      ! Note qclr<=qv<=qvs so qclr-qvs should be negative.
+      ! In practice roundoff error *might* make qrevp slightly negative, so adding max(0,...)
+      ! The minus sign in front makes qrevp positive
+      qrevp = max(0._rtype, -epsr * (qclr-qvs)/ab )
```

Comment:
@PeterCaldwell 

- The rain `qrevp` depend here on 'pushing' the clear-sky area to the level of saturation (which is a large value; in reality it will be hard getting a volume of air saturated just by evaporation).

- The `epsr` factor should of course bound it. However, looking at the WRF version, this term is multiplying by a term `oxx*(1-exp(-xx*dt))`. `oxx` is the total relaxation time (cloud, rain, ice). The total relaxation time involves info about `qc,nc,qr,nr,qi,ni` and mean sizes - thus getting the CLUBB changes directly (explicit). 
My point here is that according to the current formulation, you might get a large number which still needs to be bounded (factored).

- Possibly we can discuss it offline, but referring to the `qrevp` WRF version: assuming that the best we can do is to use the temperature and water vapor tendencies (e.g., T - T_old, qv - qv_old) which stores the current time step changes in condensation/evaporation happen in CLUBB - is there additional in-consistency you this will misrepresent `qrevp`?

Reply:
Interesting point. It is hard to imagine that any parameterization of rain evaporation won't boil down to the difference between clear-sky and saturated air scaled by the supersaturation removal timescale epsr and accounting for the fact that evaporation cooling increases qsat, damping evaporative strength (ab in the parameterization above). I took epsr from the p3 table lookup without thinking about it, but perhaps this version of epsr is inconsistent with our current implementation of qvclr - qsat?

Reply:
I think the formulation for `epsr` is similar to the one done in WRF-p3.
Your `qrevp` formulation seems to lack additional scale with the actual MP time-step (`dt` or `odt`) and total relation time (`xx` or 'oxx') as per the original WRF-p3 scheme.

The formulation would have looked then like:   
`qrevp = (epsr * (qclr-qvs) * odt*oxx*(1.0-exp(-xx*dt))) / ab` 

And again, this is only a single term (out of 3) in the WRF-p3, so not sure the magnitude would be bounded ...